httpbp/headers.go

```diff
@@ -0,0 +1,188 @@
+// baseplate resource (spans and edge contexts) into a struct that provides an
+// easy way to convert them into HTTP headers.
+//
+// This interface exists so we can avoid having to do runtime checks on maps to
+// ensure that they have the right keys set when we are trying to sign or verify
+// a set of HTTP headers.
+type Headers interface {
+	// AsHTTPHeaders returns the Headers struct as HTTP headers, a map of header
+	// keys to header values.
+	AsHTTPHeaders() map[string]string
+}
+
+// EdgeContextHeaders implements the Headers interface for HTTP EdgeContext
+// headers.
+type EdgeContextHeaders struct {
+	EdgeRequest string
+}
+
+// NewEdgeContextHeaders returns a new EdgeContextHeaders object from the given
+// HTTP headers.
+func NewEdgeContextHeaders(h http.Header) *EdgeContextHeaders {
+	return &EdgeContextHeaders{
+		EdgeRequest: h.Get(edgeContextHTTPHeaderKey),
+	}
+}
+
+// AsHTTPHeaders returns the EdgeContextHeaders as HTTP headers, a map of header
+// keys to header values.
+func (s *EdgeContextHeaders) AsHTTPHeaders() map[string]string {
+	return map[string]string{
+		edgeContextHTTPHeaderKey: s.EdgeRequest,
+	}
+}
+
+// SpanHeaders implements the Headers interface for HTTP Span headers.
+type SpanHeaders struct {
+	TraceID  string
+	ParentID string
+	SpanID   string
+	Flags    string
+	Sampled  string
+}
+
+// NewSpanHeaders returns a new SpanHeaders object from the given
+// HTTP headers.
+func NewSpanHeaders(h http.Header) *SpanHeaders {
+	return &SpanHeaders{
+		TraceID:  h.Get(traceIDHTTPHeaderKey),
+		ParentID: h.Get(parentIDHTTPHeaderKey),
+		SpanID:   h.Get(spanIDHTTPHeaderKey),
+		Flags:    h.Get(spanFlagsHTTPHeaderKey),
+	}
+}
+
+// AsHTTPHeaders returns the SpanHeaders as HTTP headers, a map of header keys
+// to header values.
+func (s *SpanHeaders) AsHTTPHeaders() map[string]string {
+	return map[string]string{
+		traceIDHTTPHeaderKey:     s.TraceID,
+		parentIDHTTPHeaderKey:    s.ParentID,
+		spanIDHTTPHeaderKey:      s.SpanID,
+		spanFlagsHTTPHeaderKey:   s.Flags,
+		spanSampledHTTPHeaderKey: s.Sampled,
+	}
+}
+
+var (
+	_ Headers = (*EdgeContextHeaders)(nil)
+	_ Headers = (*SpanHeaders)(nil)
+)
+
+// PopulateBaseplateRequestContext returns a function that takes baseplate HTTP
+// headers from the request, verifies that it should trust the headers using the
+// provided HeaderTrustHandler, and attaches the trusted headers to the context.
+//
+// The function that this produces can be passed to go-kit's http.ServerBefore
+// ServerOption.
+//
+// These headers can be retrieved using httpbp.GetHeader.
+//
+// example:
+//		trustHandler := httpbp.NeverTrustHeaders{}
+//		httpgk.NewServer(
+//			endpoints.IsHealthy,
+//			DecodeIsHealthyRequest,
+//			EncodeResponse,
+//			httpgk.ServerBefore(
+//				httpbp.PopulateRequestContext(trustHandler),
+//			),
+//		)
+func PopulateBaseplateRequestContext(t HeaderTrustHandler) func(ctx context.Context, r *http.Request) context.Context {
```

Comment:
We probably should also provide the inner function in the API, like `InjectContextFromHTTPHeader` or something, because this one can only be used with go-kit http package, some microservices don't really want to be forced to use that :)